Theme: consistant theming for applet drag handles in all themes

lukefromdc opened the pull request:
*Do NOT use border, as it is not overridden in MATE themes and pushes the image out of place
*Use two gradient images to make a 3x3 grid resembling MATE theme 3 holes image that works in both vertical and horizontal panels
*Note that while drag handle now shows in vertical panel, it only scales in horizontal panel as px and not % must be used in horizontal direction

raveit65 commented:
Let's hold this for a moment.
See my comment in PR for 1.18 branch.
https://github.com/mate-desktop/mate-panel/pull/781#issuecomment-382722379
I am not really sure about
*MatePanelAppletFrameDBus > MatePanelAppletFrameDBus { }*

lukefromdc commented:
I just tested the switch to just using MatePanelAppletFrameDBus and got ugly results on in-process applets:

lukefromdc commented:
Force-pushed branch after fixing an indent error

lukefromdc commented:
Should I go ahead and merge this, having established the need for the double selector?

lukefromdc commented:
I did a retest with out of process applets, they looked the same as with in-process so no trouble on that front

raveit65 commented:
Yeah, merge it.
I am guessing we need the same for 1.18 branch for gtk+-3.22/20, or not ?

monsta commented:
Hmm, dots now look properly aligned with Menta, TraditionalGreen, BlackMATE... but there are no dots/lines at all with Adwaita, Adwaita Dark, HighContrast, HighContrastInverse.

I'm using mate-themes from git master, so I have https://github.com/mate-desktop/mate-themes/commit/10886a79f01abf33f55de321b5ff96d9ada0bbd9 and https://github.com/mate-desktop/mate-themes/commit/cf0a970f26d30c5db7c0af25396eb6e2dc1924a0. I don't have any in-process applets. GTK+3 version is 3.22.30.

monsta commented:
Just tried git master once again to make sure... Both Adwaita themes have dashes, though not very good looking (same as on screenshots in https://github.com/mate-desktop/mate-panel/pull/781), and both HighContrast themes have large dots.

muktupavels commented:
Based on pull request title? Why not use symbolic image/icon for panel handle? And than use css to only control/adjust color?

Also why you don't use `background-size` to control gradient size? For example `background-size: 14px 22px` and then with `background-position` you can control where it appears... That why panel size would not change how it looks.

How often `mate_panel_applet_frame_dbus_change_background` is called? You every time create new css provider. You do understand that GTK does not automatically remove your previous providers, right? So if that function has called 100 times you have added 100 providers, possibly for different themes with same priority...

monsta commented:
TraditionalGreen (dots ok):
![traditionalgreen](https://user-images.githubusercontent.com/5138986/39045719-504a3b82-449c-11e8-986d-d66c6ef546ce.png)

Adwaita:
![adwaita](https://user-images.githubusercontent.com/5138986/39045730-615e4418-449c-11e8-8971-98b718d28536.png)

Adwaita Dark:
![adwaita-dark](https://user-images.githubusercontent.com/5138986/39045740-69d8e508-449c-11e8-9225-b782cc71b2f9.png)

HighContrast:
![highcontrast](https://user-images.githubusercontent.com/5138986/39045749-745edc12-449c-11e8-9b10-0a69a9d5f306.png)

HighContrastInverse:
![highcontrastinverse](https://user-images.githubusercontent.com/5138986/39045756-7fa426c2-449c-11e8-9d57-5a1d094bbbf9.png)

monsta commented:
> you have tested with mate-themes git master ?

Yes

@alexarnaud: can you please test it too?

alexarnaud commented:
Le 20/04/2018 à 12:29, monsta a écrit :
> Yes
> 
> @alexarnaud <https://github.com/alexarnaud>: can you please test it too?
> 

Yes, I can, the test you're requested to me is only on master, not 1.18, 
right?

Best regards,
Alex.

raveit65 commented:
> (I don't have installed adwaita to test)

@sc0w 
Adwaita and adwaita-dark is always installed if gtk+ is installed.
![m-a-p](https://user-images.githubusercontent.com/961604/39052628-f48bee9c-44ac-11e8-8f7c-ab89d3f61a8f.png)
Add a metacity theme to their index.them file to use a thumbnail. Eg.
/usr/share/themes/Adwaita-dark/index.theme
```
GtkTheme=Adwaita-dark
IconTheme=Adwaita
MetacityTheme=Adwaita-dark
CursorTheme=Adwaita
CursorSize=24
```

raveit65 commented:
Ok, for gtk+ themes the Handle is there for na-tray applet but not for the windows-list-applet.
Sorry , i always looked at the na-tray applet for testing.

lukefromdc commented:
OK, I have one other choice and will open an alternate PR, removing the offending border and doing nothing else. I tried to load a GtkIcon but that cannot be done, it did not show up and I got runtime warnings that MatePanelAppletFrameDBus is a subclass of GtkBin and can only accept one widget, thus excluding any GtkImage etc.

The tray looks fine in my install. It is possible we may need to move the CSS provider to the initial start of the applet to prevent repeated calls as Albert said. I DID test using background-size, but that can clip the default MATE image as it is not removed in MATE themes. If the MATE themes cannot be adjusted to remove anything added in that CSSprovider (as the image itself already gets replaced), than this gets really complicated really fast.

lukefromdc commented:
Alternate PR https://github.com/mate-desktop/mate-panel/pull/784
fixes the image being cut in mate themes without changing anything else. Not as pretty in Adwaita, Adwaita-dark etc though no changes in the high contrast themes as they are special-cased. The mate themes now look right, and with the default image going all the way to the left there should never be a problem with the window list (though the window list looked the same as the tray in my previous PR as well).


![Uploading Adwaita-Dark.png…]()

lukefromdc commented:
@monsta , what are the A11y issues mentioned, and what change are you requesting? Also check out the alternate PR https://github.com/mate-desktop/mate-panel/pull/784 which is simpler but a bit ugly in Adwaita and Adwaita-Dark

lukefromdc commented:
Why not use symbolic image/icon for panel handle? Because you can't load two widgets into a GtkBin,of which MatePanelAppletFrameDBus is a subclass. Thus a rewrite of the whole widget and all of the themes would be needed to do something like that. What I need is a way to load an image file: easy enough if we could assume the image would unconditionally be in /usr/share/mate-panel, but that's now how build and install options work. I don't see any way in the world to get a path that contains a variable into a cssprovider  and loading a GtkImage does not work because of the GtkBin issue.

Meanwhile this is greatly complicated by competing needs: Raleigh and win32 do not use the @theme_bg_color and instead use @bg_color (etc) selectors so require one special case if the colors are to vary with themes. The high contrast themes require black or white if there is to in fact be high contrast. A border cuts the MATE themes. I have no idea in the world what the referenced A11y issue is, and thus no starting point to address it.

I could reduce to horizontal component of the image in this PR to replace the 3X3 grid(meant to help the vertical panel case) with just a 3 hole vertical image, maybe that would help some of the concerns raised here, or not?  It would shorten the code a bit, and vertical panel support in mate-themes will require a rework anyway for this item.

I know of exactly ONE simple approach: go all the way back to the original system,. where I only set a background color and nothing else. That uses either a solid color (if not transparent) or if an alpha is used the old GTK default mesh appearance. We lose the MATE 3 holes image entirely in GNOME themes but the handle stays visible, and we can still special-case the high contrast themes

muktupavels commented:
You can use symbolic icon / image with `background-image` and control its color with `color`...
You can create css provider once and reuse/change css in it when needed...
You can create new providers, but then remove old providers that are not needed...
You can get event/callback when theme changes...
You can use GResource to ship icon/image with/in mate-panel so it is always available...

I would special case only gtk+ builtin themes. Your own themes probably already works good, no? Then why complicate things and add css that could do bad things to existing themes? Other themes can fix their themes if they want support MATE.

lukefromdc commented:
As a test, I was able to load a .svg image file from an absolute path (cannot ship that way though) but was not able to recolor it. Even borrowing a symbolic icon and calling it grig-symbolic.svg it was always renderede in its original color

muktupavels commented:
```
background-image:-gtk-recolor(url('/usr/share/icons/Adwaita/scalable/actions/action-unavailable-symbolic.svg'));
color: #ff0000;
```

lukefromdc commented:
That works, and it turns out our existing panel-grid.svg can function as a symbolic image. Best of all, I was able to strip out all code for special-casing the high contrast themes, as they now get the foreground color. Works in all GNOME themes except in win32, where the image still appears but is very faint as @theme_fg_color is not used in win32. This will work with ANY theme that uses @theme_fg_color, and as the background is transparent does not write over the panel's existing background or need @theme_bg_color which is not present in Raleigh.

Here is the cssprovider used, it is much simpler and could be moved to startup:
```

    GtkCssProvider *provider;
    provider = gtk_css_provider_new ();
     {
	    gtk_css_provider_load_from_data (provider,
            "MatePanelAppletFrameDBus > MatePanelAppletFrameDBus { \n"
            "border-style: none; \n"
            "background-repeat: no-repeat; \n"
            "background-position: left; \n"
            "color: @theme_fg_color;"
            "background-image: -gtk-recolor(url(\"/usr/share/mate-panel/grid-symbolic.svg\")) \n"
            "}",
            -1, NULL);
        }
	gtk_style_context_add_provider (gtk_widget_get_style_context(GTK_WIDGET(frame)),
                                    GTK_STYLE_PROVIDER (provider),
                                    GTK_STYLE_PROVIDER_PRIORITY_FALLBACK);
    g_object_unref (provider);

```

lukefromdc commented:
https://github.com/mate-desktop/mate-panel/tree/dev-applet-handle-wip
requires dropping  https://github.com/mate-desktop/mate-panel/files/1933891/panel-grid-symbolic.svg.gz into /usr/share/mate-panel to work (it is a WIP needing more work) but can be tested with that image file. What I now need is a way to install that image, either as a gresource or any other way that works now matter how mate-panel is configured as to installation directories

lukefromdc commented:
I finally got an image to work and install properly! I realized that a css file loaded from a path could refer to an image file by relative path, same as in a theme file. Putting mate-panel.css and panel-grid-symbolic into the data directory installed them to /usr/share/mate-panel with my build options, and they worked. I moved the CSS loading into mate_panel_applet_frame_dbus_init so it would be loaded once only.  MUCH simpler code and works better too.

On changing themes the color changes fine, as the same css file works with whatever colors the themes feed them. Only in the win32 theme is there an issue as that theme does not use @theme_fg_color,   so there the image still shows up, it just is not recolored and looks a lot like it does in Menta exept for being slightly lighter rather than slightly darker. Solid contrast in the high-contrast themes, as they set background and foreground colors opposite.

Now we get the mate-panel drag handle dots from the mate-themes on the drag handles in all the GNOME themes and any other themes not supporting the panel. For the mate-themes we can ship the image (it's the same file) for backward compatability or not, and if desired can recolor it to anything desired whether it is the default image or the image from the theme file, it should make no difference at all.

Rebased for the latest commit to master, please retest

raveit65 commented:
https://git.gnome.org/browse/gnome-panel/log/?qt=grep&q=gresource
And there is a later commit where @muktupavels added the gnome-panel.css

raveit65 commented:
Looks good with our themes.
But with adwaita or any other theme which doesn't support mate-panel.
![windows-list-applet-handle](https://user-images.githubusercontent.com/961604/39082781-ecc794f2-4558-11e8-8479-f8fec5b89148.png)

Changing mate-panel.css to 
*background-image: -gtk-scaled(url("panel-grid-symbolic.svg"))*
fixes the problem here.
![windows-list-applet-handle-good](https://user-images.githubusercontent.com/961604/39082798-15c6f442-4559-11e8-9ddf-ef858b2a5d99.png)
This is the same what we use in themes css.

PS: i will add the gresource stuff when all is done, if you agree.

raveit65 commented:
More testing points out the *-gtk-recolor* term is needed. But we need to set up a fixed background-size for HIDPI.
```
MatePanelAppletFrameDBus > MatePanelAppletFrameDBus {
	border-style: none;
	background-repeat: no-repeat;
	background-position: left;
	color: @theme_fg_color;
	background-image: -gtk-recolor(url("panel-grid-symbolic.svg"));
	background-size: 12px 22px;
}
```
This works for me with all themes.
Eg. HCI theme.
![hci](https://user-images.githubusercontent.com/961604/39083409-9157f3fe-4563-11e8-8b4e-7b7186f4398f.png)

raveit65 commented on data/mate-panel.css:
```diff
@@ -0,0 +1,9 @@
+	background-repeat: no-repeat;
+	background-position: left;
+	color: @theme_fg_color;
+	background-image: -gtk-recolor(url("panel-grid-symbolic.svg")) 
```
small typo, missing semicolon.

lukefromdc commented:
I just fixed the mate-panel.css file, adding the missing semicolon and the fixed size. Works nicely here, force-pushed to avoid further complicating the PR

lukefromdc commented:
Also, switching this to gresource is fine, makes the code a bit lighter at runtime so a good idea

muktupavels commented:
I would add global css provider not create new provider for each applet with same css.

raveit65 commented:
Can we also squash both commits  before merging?

lukefromdc commented:
OK, this is squashed

lukefromdc commented:
Should I go ahead and merge or are you adding the gresource conversion to the same PR,  for which we would need to wait

lukefromdc commented:
@monsta , please retest. Does this fix the issues you found?

lukefromdc commented:
Just fixed an indent error in mate-panel.css

lukefromdc commented:
I just switched the cssprovider over to loading once only in main.c for the whole panel, same as it does in Caja

raveit65 commented:
I just tested it again and it works like expected.
I will merge it as it blocks my local repo with adding GResource to mat-panel.

muktupavels commented on mate-panel/main.c:
```diff
@@ -60,7 +60,8 @@ main (int argc, char **argv)
 	char           *desktopfile;
 	GOptionContext *context;
-	GError         *error;
+	GError         *error, *error2;
```
Why second GError?

monsta commented:
@lukefromdc: I mentioned a11y issues because dots were lost for both high contrast themes, therefore it became the same as if https://github.com/mate-desktop/mate-panel/pull/779 wasn't applied. Now it looks alright, I just built from master, and dots are shown in all themes.

alexarnaud commented:
@raveit65 Do you think we need more backports for 1.18 or 1.20 branches ?

Best regards,
Alex.

raveit65 commented:
@alexarnaud 
already done for 1.20
backport and testing for 1.18 should be done by @monsta 
I don't update f26 with 1.18 any more as it is EOL in 1.5 months.

lukefromdc commented:
The second gerror was used because I was unsure whether re-using the first one would work or would be a potential source of problems. The css block comes after a g_error_free (error) statement for the first gerror variable.

muktupavels commented:
If `g_option_context_parse` fails, `mate-panel` exits, no? So how would you expect it to reach css part?

Even if it would not be fatal error and `mate-panel` would continue to run you would just have to reset it to `NULL` after freeing it.

But have you read documentation? You are supposed to connect to `parsing-error` signal to track errors in css loading.

monsta commented:
@raveit65: yeah, I'll backport it to 1.18, just need to limit it to GTK+ 3.20/22, same as done in https://github.com/mate-desktop/mate-panel/commit/3a8b1f66d2393a35f40dbcc211d3bf045d563a92.

raveit65 commented:
By the way, should not the old malfunctioning handle code be removed?
https://github.com/mate-desktop/mate-panel/blob/master/mate-panel/panel-applet-frame.c#L100

alexarnaud commented:
@monsta could you please tell us here when you've pushed to 1.18? I'm waiting the backport to provide new mate-panel 1.18 to a visual-impaired user that uses HighContrast.

Thank you very much @lukefromdc @monsta @lukefromdc @muktupavels for your involvement in this area :) !

Best regards,
Alex.

monsta commented:
Sure, I'll post here.

monsta commented:
Backported to 1.18 in https://github.com/mate-desktop/mate-panel/commit/4437af31ff82629c18f53f6f4c38430547378af9.

lukefromdc commented:
It looks like the only thing that broke in the old code at and below
https://github.com/mate-desktop/mate-panel/blob/master/mate-panel/panel-applet-frame.c#L100
was gtk_render_handle, which will only display the old mesh background at least in GTK 3.22 if a background color is explicitly set. It is still needed because if it is removed, the new image won't display either. The rest of the code displays the background for the applet outside the handle, and removing it causes any theme that does set a bg color for the handle to render it over the entire applet and not just the handle. Thus it also cannot be removed.

Only the part within the 

```
	if (bg_pattern && (background->type == PANEL_BACK_IMAGE ||
	    (background->type == PANEL_BACK_COLOR && background->has_alpha)))
```

if block seems to have no effect and may be something we can remove. I will try to fix the custom background cases

lukefromdc commented:
On further experiment no success with getting the image into applet handles with a custom background, I think .mate-custom-panel-background is overriding it, and specifying MatePanelAppletFrameDBus.mate-custom-panel-background as well in the css file (the experimental single should cover the double selector case) does nothing. Probably the removal of backgrounds elsewhere to allow custom backgrounds to show is still catching it.

muktupavels commented:
>  It is still needed because if it is removed, the new image won't display either. 

Why not look what gtk is actually doing in `gtk_render_handle`? If you would have done that then you probably would have find that you should be able to replace it with `gtk_render_background` and `gtk_render_frame`.

lukefromdc commented:
gtk_render_handle is not actually deprecated, not sure why the original patterns stopped rendering without an explicit background color set, but I really don't have time to dig into that now that we no longer use it. No mate-theme and no GNOME theme sets a transparent or shaded background for the handle, the condition I've found makes it work. Themers can still do that if they want, and can override and remove the new default image. Thus, this is not broken, the old behavior is just undocumented..

Meanwhile, we've got a lot of other stuff to work on
